### PR TITLE
VC slashing Part 8 - vc slashing at

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
Add validator slashing acceptance tests

## Fixed Issue(s)

#7707 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (zilm13) [inline acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: It would be more readable if you extract it to something like `  private SignedBeaconBlockHeader randomSignedBeaconBlockHeader(final UInt64 slot,final UInt64 validatorIndex,
final SigningRootUtil signingRootUtil, final ForkInfo forkInfo, final BLSSecretKey secretKey) ` and use 2 times

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (zilm13) [inline acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: Same here, please extract random indexedAttestation generation to private method

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: I think there should be added ` tekuNode.waitForExit();` to this test and all others. I've added it and it doesn't work. Any ideas?

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: nit: I'd put `tekuNode`  to the top of the list as it's most crucial value for the test

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: it would be good to see something like `int slotInThirdEpoch = firstTekuNode.getSpec().forMilestone(SpecMilestone.ALTAIR).getSlotsPerEpoch() * 2 + 3;` above for clarity of this magic number

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: I don't understand why it's `NoBlocks` case, both nodes will have all blocks in this case. Protection activation could be the result of either gossip or block in this case.

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: I don't understand why it's `NoBlocks` case, both nodes will have all blocks in this case. Protection activation could be the result of either gossip or block in this case.

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: It would be good to have a description here why we need to test so many cases (single process bn+vc/ separate vc, block slashing/gossip slashing)

**Review by zilm13 — COMMENTED** [review]
I'm not sure we have a way to cover NoBlock case separately. But maybe I'm wrong. If not I'd just remove these cases.
Also it would be great to see in the end of each test something like
```
secondTekuNode.waitForExit();
firstTekuNode.waitForSlot();
```
to see that correcct node is shutdown and everything else not

lines like `firstTekuNode.stop();` are not needed, it should be shutdown automatically

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: My bad, forgot to clean up this mess 🙈

Comment (mehdi-aouadi) [reply acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: Done

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: Done

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done for all

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done

Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Let me know if that's what you had in mind

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: But there's a param which is the validator key and it's not a simple exit but rather a specific slashing shut down log... Not sure I get what you mean here

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: This is "no blocks" because the first node is not running any validator and hence won't produce any block. The slashing post is sent to that same node which is then gossiped...

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Same here. In the "no blocks" scenarios we're simply not running validators and assuming that node won't produce any block...

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: But is there any difference when node is block producer or just received block via gossip for this protection?

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: But the process exits?
I mean I'd be more happy to know that it's not only logline but Teku actually shutdown

Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Oh I see what you mean. Turned out to be a timeout issue (the existing `waitForExit` waits 30s and that wasn't enough), fixed it by adding a new `waitForShutDown`. I also added some checks to make sure the non concerned nodes are not shut down in there own...

**Comment by mehdi-aouadi** [comment]
> I'm not sure we have a way to cover NoBlock case separately. But maybe I'm wrong. If not I'd just remove these cases. Also it would be great to see in the end of each test something like
> 
> ```
> secondTekuNode.waitForExit();
> firstTekuNode.waitForSlot();
> ```
> 
> to see that correcct node is shutdown and everything else not
> 
> lines like `firstTekuNode.stop();` are not needed, it should be shutdown automatically

Should be better now. Removed some non required manual shut downs and added some checks to make sure only the slashed nodes are shut down

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: If we'd like to simulate the slashing being sent within a block and not the slashing gossip topics, the node which receives the slashing should produce a block in order to include the slashing in it.
It goes like the following, assuming the slashed validator is run by Node 2
- Block scenario: Post slashing to Node 1 -> Node 1 produces a block including the slashing -> Node 1 gossip the block -> Node 2 receives the blocks and sees the slashing -> Node 2 shut down
- Proposer slashing or attester slashing gossip messages: Post slashing to Node 1 -> Node 1 sends the slashing message to the proposer or attester slashing gossip topic -> Node 2 receives the proposer or attester slashing message through gossip -> Node 2 shut down

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (zilm13) [inline acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: nit: you could remove 1's here and there

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: SSE

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/Node.java
Comment (zilm13) [inline acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/Node.java]: it's very unclear difference in naming, I'd rather call it the same `waitForExit` but with `seconds` parameter leaving a comment on standard one that it will wait for 30 seconds, which could be not enough for complete node shutdown

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: and SSE

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (zilm13) [inline acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: final, please

**Review by zilm13 — COMMENTED** [review]
LGTM, but let's wait for approval until APAC feedback on #7954
Also let's split it for clarity to `ValidatorSlashingDetectionAcceptanceTest` and `ValidatorSlashingDetectionNightlyAcceptanceTest`, so class description will need split too with nightly link. All tests in nightly should be disabled. What tests do you think we should leave in this one? I'd leave just first two

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java]: My bad. Done

File: acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/Node.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/Node.java]: Done

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done

File: acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java
Comment (mehdi-aouadi) [reply acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorSlashingDetectionAcceptanceTest.java]: Done

**Comment by mehdi-aouadi** [comment]
> LGTM, but let's wait for approval until APAC feedback on #7954 Also let's split it for clarity to `ValidatorSlashingDetectionAcceptanceTest` and `ValidatorSlashingDetectionNightlyAcceptanceTest`, so class description will need split too with nightly link. All tests in nightly should be disabled. What tests do you think we should leave in this one? I'd leave just first two

IMO we could keep the two single peer tests (standalone VC and single process VC/BN: `shouldShutDownWhenOwnedValidatorSlashed_SingleProcess_SinglePeer` and `shouldShutDownWhenOwnedValidatorSlashed_StandAloneVC_SinglePeer`) which are pretty fast and broadly cover the feature. The other ones that cover more complex combinations/scenarios could be left to the nightly run if approved...

**Review by zilm13 — APPROVED** [review]
LGTM 
just a package name way too long even for Java
how about `slashingprotection`?
and we could also move Doppelganger test there too

**Comment by mehdi-aouadi** [comment]
> LGTM just a package name way too long even for Java how about `slashingprotection`? and we could also move Doppelganger test there too

Renamed the package to `validatorslashing` and simplified the tests class names too.
I left the doppelganger where it is since it's a bit different topic...

**Pull request merged**
